### PR TITLE
Consolidate Using Puppet for Config. Management

### PR DESCRIPTION
**Pull request by Lennonka**
This is the first step to remaking the "Using Puppet" chapter in the _Managing Hosts Guide_ into a future standalone upstream _Puppet Guide_.

This is an intermediate improvement that involves:

- Remake of the chapter ToC structure
- Rephrasing of some headings
- Modularization of source code

I changed content minimally, only for it to make sense with the new headings.

Do not cherry-pick.

### CODE REVIEW

**Comment by maximiliankolb** [comment]
I've talked to @nadjaheitmann about Puppet intervals. She told me she might work on this again soon (maybe even this week), and then re-review the current documentation on that and especially its heading.

Ping me once your draft is ready for review. One thing I already notices is `assembly` in anchors, which we generally don't include. But :+1: for excluding headings from assemblies.

**Comment by Lennonka** [comment]
@maximiliankolb Thank you, that's great news. I wouldn't sweat it, we can and will update headings in the next round again. Will fix the IDs.

File: guides/common/modules/proc_configuring-a-puppet-class.adoc
Comment (ekohl) [inline guides/common/modules/proc_configuring-a-puppet-class.adoc:3]: Yes. Puppet modules have built-in defaults but within the application you can override those defaults. If you want to override a parameter within the application (globally, on a host group or on hosts), it must be marked as overrideable.

One thing this doesn't mention (perhaps it's mentioned elsewhere) is that you can also mark a parameter as overrideable but then not override the built-in default. There is a checkbox `Omit` to do so. IMHO that should be the default, but the documentation should at the very least mention it.

File: guides/common/modules/ref_customizing-the-registration-templates.adoc
Comment (ekohl) [inline guides/common/modules/ref_customizing-the-registration-templates.adoc:39]: Perhaps it should say:
```suggestion
* The `puppetlabs_repo` and `puppet_setup` snippets downloads and installs Puppet agent on the host (only when a Puppet server is assigned)
```

My reasoning is that a Puppet server being present is ambiguous. Only when a host has a Puppetserver assigned, it'll be used. This means you can have hosts with and without Puppet in the same environment.

**Review by ekohl — COMMENTED** [review]
I haven't read through it too closely, just some things that jumped out to me.

File: guides/common/modules/ref_customizing-the-registration-templates.adoc
Comment (Lennonka) [reply guides/common/modules/ref_customizing-the-registration-templates.adoc:39]: Perfect, thanks. This felt unclear when I read it.

File: guides/common/modules/proc_configuring-a-puppet-class.adoc
Comment (Lennonka) [reply guides/common/modules/proc_configuring-a-puppet-class.adoc:3]: @ekohl I can add it, but I would prefer to know in which use cases is benefical to only mark a parameter overridable without overriding its value.

File: guides/common/modules/ref_customizing-the-registration-templates.adoc
Comment (ekohl) [reply guides/common/modules/ref_customizing-the-registration-templates.adoc:39]: Also, should it be plural?
```suggestion
* The `puppetlabs_repo` and `puppet_setup` snippets download and install Puppet agent on the host (only when a Puppet server is assigned)
```

File: guides/common/modules/proc_configuring-a-puppet-class.adoc
Comment (ekohl) [reply guides/common/modules/proc_configuring-a-puppet-class.adoc:3]: If you want to use the built in default from the Puppet manifest. Let's say you have:
```puppet
class myclass (
  String $servername = $facts['fqdn'],
) {
  # ...
}
```

This essentially means you have a class named `myclass`. If you import it, it will find the parameter `servername` with a default `$facts['fqdn']`. If you click on `override`, that literal string is copied into Foreman's default field.

To really get an equivalent default, you would need to use some ERB macro that I don't even know from the top of my head. Something like `<%= ... %>` sould be the default in Foreman's default field. If you use `omit`, you can continue using the built-in Puppet default.

That is to me, a Puppet user, a way more natural way. Only if I want to override the built-in default with a new global default should I provide a default.

Also note that leaving the default field empty is AFAIK the empty string. Selecting `Omit` is fundamentally different.

If you have an environment where you can play, I'd recommend to try it out. If you go to the host detail page you should see a button YAML (though I think it's renamed when it was extract to a plugin). There you should see the assigned class with overrides. With `Omit`, it should not show up unless overridden at the host level. Without it will always show up.

Comment (Lennonka) [reply guides/common/modules/proc_configuring-a-puppet-class.adoc:3]: @ekohl You know, this is a great remark but it's not what this PR is concerned with. So I suggest that I will create an issue in our Jira for future work, because we will get to some rewrites of the content later. Agreed?
With this comment I was merely wondering if a different phrasing of the heading would be more suitable.

Comment (ekohl) [reply guides/common/modules/proc_configuring-a-puppet-class.adoc:3]: That's completely OK, but at least you now have a bit more understanding of the background.

Comment (Lennonka) [reply guides/common/modules/proc_configuring-a-puppet-class.adoc:3]: Absolutely and I appreciate it!

**Comment by Lennonka** [comment]
Okay, we've clarified with @ekohl. Now we're waiting for a review by @vsedmik

File: guides/common/modules/proc_setting-the-puppet-out-of-sync-interval.adoc
Comment (vsedmik) [inline guides/common/modules/proc_setting-the-puppet-out-of-sync-interval.adoc:5]:
```suggestion
. In the {ProjectWebUI}, navigate to *Administer > Settings*, and click the *Config Management* tab.
```
On downstream Satellite I can find it on *Config Management* tab.

**Review by vsedmik — CHANGES_REQUESTED** [review]
I have finally built the doc and got oriented in the structure changes provided by this PR. I think it's good improvement providing logical flow.
I have added only few proposals for change but overall it looks good to me. :+1:

File: guides/common/modules/proc_creating-a-puppet-environment.adoc
Comment (Lennonka) [reply guides/common/modules/proc_creating-a-puppet-environment.adoc:2]: @vsedmik An import of what and to where? And the environment deleted from where?

Comment (Lennonka) [reply guides/common/modules/proc_creating-a-puppet-environment.adoc:2]: I will include it before the Procedure, because it might get overlooked after it.

Comment (vsedmik) [reply guides/common/modules/proc_creating-a-puppet-environment.adoc:2]: Import of puppet environment in https://<SATELLITE>/foreman_puppet/environments

**Comment by Lennonka** [comment]
So, we should be ready for the last review before merge :) *fingers crossed*

**Comment by maximiliankolb** [comment]
> So, we should be ready for the last review before merge :) (crossing fingers) @maximiliankolb

I will review this today!

**Review by maximiliankolb — COMMENTED** [review]
Thanks for improving the puppet docs!

I personally don't like splitting the assemblies, but it's ok for me. For me, handling a long list of includes in 'master.adoc'/or an assembly is easier.

Made some comments and suggestions.

File: guides/common/modules/proc_installing-a-puppet-module.adoc
Comment (Lennonka) [reply guides/common/modules/proc_installing-a-puppet-module.adoc:2]: I noticed, but "server" is lower case in that attribute and we need uppper case for the heading.

File: guides/common/modules/proc_setting-the-global-out-of-sync-interval.adoc
Comment (Lennonka) [reply guides/common/modules/proc_setting-the-global-out-of-sync-interval.adoc:9]: Deffo a good idea.

File: guides/common/modules/proc_creating-a-puppet-environment.adoc
Comment (Lennonka) [reply guides/common/modules/proc_creating-a-puppet-environment.adoc:14]: This is also definitely valid for Satellite. I'll check others.

Comment (Lennonka) [reply guides/common/modules/proc_creating-a-puppet-environment.adoc:14]: Okay, I'm assuming it's to be excluded for orcharhino only. I verified with a SME that it applies to all the other products.

File: guides/common/modules/proc_editing-a-puppet-class.adoc
Comment (Lennonka) [reply guides/common/modules/proc_editing-a-puppet-class.adoc]: I'm suggesting removing the file altogether instead.

**Comment by melcorr** [comment]
Hey Lena, I've read all your changes but with the build broken, I cannot proceed with verifying this fully. I'll try to get to the bottom of this.

File: guides/common/modules/proc_setting-the-global-out-of-sync-interval.adoc
Comment (melcorr) [inline guides/common/modules/proc_setting-the-global-out-of-sync-interval.adoc:7]:
```suggestion
Set a duration, in minutes, after which hosts are considered to be out of sync.
```

File: guides/common/modules/proc_setting-the-puppet-out-of-sync-interval.adoc
Comment (melcorr) [inline guides/common/modules/proc_setting-the-puppet-out-of-sync-interval.adoc:7]:
```suggestion
Set a duration, in minutes, after which hosts reporting using Puppet are considered to be out of sync.
```

**Review by melcorr — CHANGES_REQUESTED** [review]
One tiny change in two places. 
I've spotted a few more things reading the guide, but not introduced by you. 
All trivial to be honest. I will use them with Akshay for git and PR training etc.

File: guides/common/modules/proc_setting-the-puppet-out-of-sync-interval.adoc
Comment (melcorr) [reply guides/common/modules/proc_setting-the-puppet-out-of-sync-interval.adoc:7]: Maybe this line could be deleted altogether? Looks duplicated?

**Review by melcorr — APPROVED** [review]
LGTM!

**Comment by melcorr** [comment]
No Cherry Pick @Lennonka ? 
Will you not need this for the Satellite 7 release?
Or are you working ahead?

**Comment by Lennonka** [comment]
@melcorr Working ahead, this is one of 3 PRs I have planned for the Testathon. If you would like this change to be in 3.1, feel free to cherrypick!

**Comment by melcorr** [comment]
@Lennonka Satellite 7 Testhathlon means 3.1 branch for sure. 
I have cherry-picked  4279b108..bc52fa1d  3.1 -> 3.1

**Comment by maximiliankolb** [comment]
Great job @Lennonka

**Pull request merged**
